rfd/00xx-snowflake-support.md

```diff
@@ -0,0 +1,170 @@
+generated by Teleport DB service. Teleport should use Database CA to sign JWT at that point to sign the request instead 
+of User CA to be able to authenticate with the key generated by `tctl auth sign` and imported to Snowflake. 
+In this way the real token is not accessible by a user and Teleport can still validate if the request 
+comes from a trusted source.
+
+```mermaid
+sequenceDiagram
+    participant client
+    participant Teleport
+    participant Snowflake
+    client->>Teleport: HTTP /session/v1/login-request
+    Teleport->>Teleport: RBAC check, genererate and replace JWT
+    Teleport->>Snowflake: HTTP /session/v1/login-request
+    Snowflake->>Teleport: Authentication Token
+    Teleport->>Teleport: Generate Teleport Auth token
+    Teleport->>client: Send Teleport Auth token
+```
+
+Nn each request Teleport can replace the "Teleport Auth token" with Snowflake Auth token:
+
+```mermaid
+sequenceDiagram
+    participant client
+    participant Teleport
+    participant Snowflake
+    client->>Teleport: HTTP /queries/v1/query-request
+    Teleport->>Teleport: Replace Auth header
+    Teleport->>Snowflake: HTTP /queries/v1/query-request
+    Snowflake->>Teleport: Response
+    Teleport->>client: Forward the response
+```
+
+With this approach the real session token never leaves the Teleport and any client is not able to bypass Teleport
+by extracting the session token.
+Snowflake session token needs to be refreshed every X minutes. For that purpose Teleport would have to internally refresh
+the token as client doesn't have access to the token (login response contains the time for how long the token is valid).
+
+#### Query logging
+
+Queries can be extracted from requests sent to `/queries/v1/query-request` endpoint, field `data.sqltext`.
+
+#### User authentication
+
+Username can be extracted from `/session/v1/login-request` requests, field `data.LOGIN_NAME`. The extracted username
+can be then use to authenticate the user against the Teleport RBAC.
+
+### CLI support 
+
+`snowsql` is the official Snowflake CLI tool, and it doesn't support TLS terminated proxies as described 
+in https://docs.snowflake.com/en/user-guide/snowsql-start.html#using-a-proxy-server, and it uses plain text communication 
+when connecting to the Snowflake on a different port that 443. 
+
+`snowsql` requires to use one of the available authentication method. By default, it asks for password unless other 
+authentication option is specified. The other alternatives are SSO and JWT/Private key (https://docs.snowflake.com/en/user-guide/snowsql-start.html).
+Teleport could use JWT/Private key for authentication. In this flow `tsh` would export user's private in PKCS8 format (at that's the only one supported by `snowsql`),
+and `snowsql` could use it to sign the JWT token.
+
+I think there are two ways on how we can support the `snowsql`: 
+
+1. `snowcli` can connect to a custom host **without** TLS on any port different from 443. 
```

Comment:
We'll have to support `tsh proxy db --tunnel` mode where the client connects to the local authenticate tunnel without TLS anyway. I would just opt for that approach for now.